File: Algorithm.Python/Alphas/MeanReversionLunchBreakAlpha.py
# QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
# Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from AlgorithmImports import *

#
# Academic research suggests that stock market participants generally place their orders at the market open and close.
# Intraday trading volume is J-Shaped, where the minimum trading volume of the day is during lunch-break. Stocks become
# more volatile as order flow is reduced and tend to mean-revert during lunch-break.
#
# This alpha aims to capture the mean-reversion effect of ETFs during lunch-break by ranking 20 ETFs
# on their return between the close of the previous day to 12:00 the day after and predicting mean-reversion
# in price during lunch-break.
#
# Source:  Lunina, V. (June 2011). The Intraday Dynamics of Stock Returns and Trading Activity: Evidence from OMXS 30 (Master's Essay, Lund University).
# Retrieved from http://lup.lub.lu.se/luur/download?func=downloadFile&recordOId=1973850&fileOId=1973852
#
# This alpha is part of the Benchmark Alpha Series created by QuantConnect which are open sourced so the community and client funds can see an example of an alpha.
#

class MeanReversionLunchBreakAlpha(QCAlgorithm):

    def initialize(self):

        self.set_start_date(2018, 1, 1)

        self.set_cash(100000)

        # Set zero transaction fees
        self.set_security_initializer(lambda security: security.set_fee_model(ConstantFeeModel(0)))

        # Use Hourly Data For Simplicity
        self.universe_settings.resolution = Resolution.HOUR
        self.set_universe_selection(CoarseFundamentalUniverseSelectionModel(self.coarse_selection_function))

        # Use MeanReversionLunchBreakAlphaModel to establish insights
        self.set_alpha(MeanReversionLunchBreakAlphaModel())

        # Equally weigh securities in portfolio, based on insights
        self.set_portfolio_construction(EqualWeightingPortfolioConstructionModel())

        # Set Immediate Execution Model
        self.set_execution(ImmediateExecutionModel())

        # Set Null Risk Management Model
        self.set_risk_management(NullRiskManagementModel())

    # Sort the data by daily dollar volume and take the top '20' ETFs
    def coarse_selection_function(self, coarse):
        sorted_by_dollar_volume = sorted(coarse, key=lambda x: x.dollar_volume, reverse=True)
        filtered = [ x.symbol for x in sorted_by_dollar_volume if not x.has_fundamental_data ]
        return filtered[:20]


class MeanReversionLunchBreakAlphaModel(AlphaModel):
    '''Uses the price return between the close of previous day to 12:00 the day after to
    predict mean-reversion of stock price during lunch break and creates direction prediction
    for insights accordingly.'''

    def __init__(self, *args, **kwargs):
        lookback = kwargs['lookback'] if 'lookback' in kwargs else 1
        self.resolution = Resolution.HOUR
        self.prediction_interval = Time.multiply(Extensions.to_time_span(self.resolution), lookback)
        self._symbol_data_by_symbol = dict()

    def update(self, algorithm, data):

        for symbol, symbol_data in self._symbol_data_by_symbol.items():
            if data.bars.contains_key(symbol):
                bar = data.bars.get(symbol)
                symbol_data.update(bar.end_time, bar.close)

        return [] if algorithm.time.hour != 12 else \
               [x.insight for x in self._symbol_data_by_symbol.values()]

    def on_securities_changed(self, algorithm, changes):
        for security in changes.removed_securities:
            self._symbol_data_by_symbol.pop(security.symbol, None)

        # Retrieve price history for all securities in the security universe
        # and update the indicators in the SymbolData object
        symbols = [x.symbol for x in changes.added_securities]
        history = algorithm.history(symbols, 1, self.resolution)
        if history.empty:
            algorithm.debug(f"No data on {algorithm.time}")
            return
        history = history.close.unstack(level = 0)

        for ticker, values in history.items():
            symbol = next((x for x in symbols if str(x) == ticker ), None)
            if symbol in self._symbol_data_by_symbol or symbol is None: continue
            self._symbol_data_by_symbol[symbol] = self.SymbolData(symbol, self.prediction_interval)
            self._symbol_data_by_symbol[symbol].update(values.index[0], values[0])


    class SymbolData:
        def __init__(self, symbol, period):
            self._symbol = symbol
            self.period = period
            # Mean value of returns for magnitude prediction
            self.mean_of_price_change = IndicatorExtensions.sma(RateOfChangePercent(1),3)
            # Price change from close price the previous day
            self.price_change = RateOfChangePercent(3)

        def update(self, time, value):
            return self.mean_of_price_change.update(time, value) and \
                   self.price_change.update(time, value)

        @property
        def insight(self):
            direction = InsightDirection.DOWN if self.price_change.current.value > 0 else InsightDirection.UP
            margnitude = abs(self.mean_of_price_change.current.value)
            return Insight.price(self._symbol, self.period, direction, margnitude, None)

File: Algorithm.CSharp/Alphas/MeanReversionLunchBreakAlpha.cs
/*
 * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
 * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

using QuantConnect.Algorithm.Framework.Alphas;
using QuantConnect.Algorithm.Framework.Execution;
using QuantConnect.Algorithm.Framework.Portfolio;
using QuantConnect.Algorithm.Framework.Risk;
using QuantConnect.Algorithm.Framework.Selection;
using QuantConnect.Data;
using QuantConnect.Data.UniverseSelection;
using QuantConnect.Indicators;
using QuantConnect.Orders.Fees;
using System;
using System.Collections.Generic;
using System.Linq;

namespace QuantConnect.Algorithm.CSharp.Alphas
{
    /// <summary>
    /// This alpha aims to capture the mean-reversion effect of ETFs during lunch-break by ranking 20 ETFs
    /// on their return between the close of the previous day to 12:00 the day after and predicting mean-reversion
    /// in price during lunch-break.
    ///
    /// Source:  Lunina, V. (June 2011). The Intraday Dynamics of Stock Returns and Trading Activity: Evidence from OMXS 30 (Master's Essay, Lund University).
    /// Retrieved from http://lup.lub.lu.se/luur/download?func=downloadFile&recordOId=1973850&fileOId=1973852
    ///
    /// This alpha is part of the Benchmark Alpha Series created by QuantConnect which are open sourced so the community and client funds can see an example of an alpha.
    ///</summary>
    public class MeanReversionLunchBreakAlpha : QCAlgorithm
    {
        public override void Initialize()
        {
            SetStartDate(2018, 1, 1);
            SetCash(100000);

            // Set zero transaction fees
            SetSecurityInitializer(security => security.FeeModel = new ConstantFeeModel(0));

            // Use Hourly Data For Simplicity
            UniverseSettings.Resolution = Resolution.Hour;
            SetUniverseSelection(new CoarseFundamentalUniverseSelectionModel(CoarseSelectionFunction));

            // Use MeanReversionLunchBreakAlphaModel to establish insights
            SetAlpha(new MeanReversionLunchBreakAlphaModel());

            // Equally weigh securities in portfolio, based on insights
            SetPortfolioConstruction(new EqualWeightingPortfolioConstructionModel());

            // Set Immediate Execution Model
            SetExecution(new ImmediateExecutionModel());

            // Set Null Risk Management Model
            SetRiskManagement(new NullRiskManagementModel());
        }

        /// <summary>
        /// Sort the data by daily dollar volume and take the top '20' ETFs
        /// </summary>
        private IEnumerable<Symbol> CoarseSelectionFunction(IEnumerable<CoarseFundamental> coarse)
        {
            return (from cf in coarse
                    where !cf.HasFundamentalData
                    orderby cf.DollarVolume descending
                    select cf.Symbol).Take(20);
        }

        /// <summary>
        /// Uses the price return between the close of previous day to 12:00 the day after to
        /// predict mean-reversion of stock price during lunch break and creates direction prediction
        /// for insights accordingly.
        /// </summary>
        private class MeanReversionLunchBreakAlphaModel : AlphaModel
        {
            private const Resolution _resolution = Resolution.Hour;
            private readonly TimeSpan _predictionInterval;
            private readonly Dictionary<Symbol, SymbolData> _symbolDataBySymbol;

            public MeanReversionLunchBreakAlphaModel(int lookback = 1)
            {
                _predictionInterval = _resolution.ToTimeSpan().Multiply(lookback);
                _symbolDataBySymbol = new Dictionary<Symbol, SymbolData>();
            }

            public override IEnumerable<Insight> Update(QCAlgorithm algorithm, Slice data)
            {
                foreach (var kvp in _symbolDataBySymbol)
                {
                    if (data.Bars.ContainsKey(kvp.Key))
                    {
                        var bar = data.Bars.GetValue(kvp.Key);
                        kvp.Value.Update(bar.EndTime, bar.Close);
                    }
                }

                return algorithm.Time.Hour == 12
                    ? _symbolDataBySymbol.Select(kvp => kvp.Value.Insight)
                    : Enumerable.Empty<Insight>();
            }

            public override void OnSecuritiesChanged(QCAlgorithm algorithm, SecurityChanges changes)
            {
                foreach (var security in changes.RemovedSecurities)
                {
                    if (_symbolDataBySymbol.ContainsKey(security.Symbol))
                    {
                        _symbolDataBySymbol.Remove(security.Symbol);
                    }
                }

                // Retrieve price history for all securities in the security universe
                // and update the indicators in the SymbolData object
                var symbols = changes.AddedSecurities.Select(x => x.Symbol);
                var history = algorithm.History(symbols, 1, _resolution);
                if (symbols.Count() > 0 && history.Count() == 0)
                {
                    algorithm.Debug($"No data on {algorithm.Time}");
                }

                history.PushThrough(bar =>
                {
                    SymbolData symbolData;
                    if (!_symbolDataBySymbol.TryGetValue(bar.Symbol, out symbolData))
                    {
                        symbolData = new SymbolData(bar.Symbol, _predictionInterval);
                    }
                    symbolData.Update(bar.EndTime, bar.Price);
                    _symbolDataBySymbol[bar.Symbol] = symbolData;
                });
            }

            /// <summary>
            /// Contains data specific to a symbol required by this model
            /// </summary>
            private class SymbolData
            {
                // Mean value of returns for magnitude prediction
                private readonly SimpleMovingAverage _meanOfPriceChange = new RateOfChangePercent(1).SMA(3);
                // Price change from close price the previous day
                private readonly RateOfChangePercent _priceChange = new RateOfChangePercent(3);

                private readonly Symbol _symbol;
                private readonly TimeSpan _period;

                public Insight Insight
                {
                    get
                    {
                        // Emit "down" insight for the securities that increased in value and
                        // emit "up" insight for securities that have decreased in value
                        var direction = _priceChange > 0 ? InsightDirection.Down : InsightDirection.Up;
                        var magnitude = Convert.ToDouble(Math.Abs(_meanOfPriceChange));
                        return Insight.Price(_symbol, _period, direction, magnitude);
                    }
                }

                public SymbolData(Symbol symbol, TimeSpan period)
                {
                    _symbol = symbol;
                    _period = period;
                }

                public bool Update(DateTime time, decimal value)
                {
                    return _meanOfPriceChange.Update(time, value) &
                        _priceChange.Update(time, value);
                }
            }
        }
    }
}